clippy_lints/src/operators/float_cmp.rs

```diff
@@ -61,54 +101,105 @@ pub(crate) fn check<'tcx>(
     }
 }
 
-fn get_lint_and_message(is_local: bool, is_comparing_arrays: bool) -> (&'static rustc_lint::Lint, &'static str) {
-    if is_local {
-        (
-            FLOAT_CMP,
-            if is_comparing_arrays {
-                "strict comparison of `f32` or `f64` arrays"
-            } else {
-                "strict comparison of `f32` or `f64`"
-            },
-        )
-    } else {
-        (
-            FLOAT_CMP_CONST,
-            if is_comparing_arrays {
-                "strict comparison of `f32` or `f64` constant arrays"
-            } else {
-                "strict comparison of `f32` or `f64` constant"
-            },
-        )
-    }
-}
-
 fn is_allowed(val: &Constant<'_>) -> bool {
     match val {
         // FIXME(f16_f128): add when equality check is available on all platforms
+        Constant::Ref(val) => is_allowed(val),
         &Constant::F32(f) => f == 0.0 || f.is_infinite(),
         &Constant::F64(f) => f == 0.0 || f.is_infinite(),
-        Constant::Vec(vec) => vec.iter().all(|f| match f {
-            Constant::F32(f) => *f == 0.0 || (*f).is_infinite(),
-            Constant::F64(f) => *f == 0.0 || (*f).is_infinite(),
+        Constant::Vec(vec) => vec.iter().all(|f| match *f {
+            Constant::F32(f) => f == 0.0 || f.is_infinite(),
+            Constant::F64(f) => f == 0.0 || f.is_infinite(),
             _ => false,
         }),
+        Constant::Repeat(val, _) => match **val {
+            Constant::F32(f) => f == 0.0 || f.is_infinite(),
+            Constant::F64(f) => f == 0.0 || f.is_infinite(),
+            _ => false,
+        },
         _ => false,
     }
 }
 
-// Return true if `expr` is the result of `signum()` invoked on a float value.
-fn is_signum(cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
-    // The negation of a signum is still a signum
-    if let ExprKind::Unary(UnOp::Neg, child_expr) = expr.kind {
-        return is_signum(cx, child_expr);
+// This is a best effort guess and may have false positives and negatives.
+fn is_pure_expr<'tcx>(cx: &LateContext<'tcx>, e: &'tcx Expr<'_>) -> bool {
```

Comment:
I assume you created a custom heuristic opposed to [`Expr::can_have_side_effects`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/struct.Expr.html#method.can_have_side_effects) to have more fine-grained control over what is considered `pure`?

Reply:
`can_have_side_effects` doesn't even allow `1.0 + 1.0` so it's completely useless here. Since this is used to avoid linting and we want to be able to handle cases like `x.trunc() == x` we need to have some heuristics that aren't appropriate for most cases.